### PR TITLE
add support for Azure SQL by adding Active Directory authentication methods

### PR DESCRIPTION
**Pull request by swanderz**
rehash of #49 but to master branch now. @nandanhegde15 will add some more documentation and test this PR does not break the existing on-prem SQL Server connection methods

### CODE REVIEW

File: dbt/adapters/sqlserver/connections.py [hunk 1]
```diff
@@ -40,11 +63,17 @@ def type(self):
     def _connection_keys(self):
         # return an iterator of keys to pretty-print in 'dbt debug'
         # raise NotImplementedError
-        return 'server', 'database', 'schema', 'port', 'UID', 'windows_login'
+        if self.windows_login is True:
+            self.authentication = "Windows Login"
+        
+    
+        return 'server', 'database', 'schema', 'port', 'UID', \
+                 'authentication', 'encrypt'
```
Comment (NandanHegde15) [inline dbt/adapters/sqlserver/connections.py:71]: So instead of displaying `windows_login` as True in `dbt debug` output , this changes would display `authentication: Windows Login`
We could have added windows auth as an Authentication value within the profiles config but
to have backward compatibility for Windows Auth with the config property : `windows_login ` otherwise `dbt debug` window would display `authentication=sql`

**Comment by NandanHegde15** [comment]
> rehash of #49 but to master branch now. @NandanHegde15 will add some more documentation and test this PR does not break the existing on-prem SQL Server connection methods

I have tested out both the Windows auth and SQL auth  part for an On Prem SQL server with this branch `Connections.py`
image

File: dbt/adapters/sqlserver/connections.py [hunk 2]
```diff
@@ -97,16 +126,55 @@ def open(cls, connection):
...
+            elif type_auth == 'sql':
+                con_str.append("Authentication=SqlPassword")
+                con_str.append(f"UID={{{credentials.UID}}}")
+                con_str.append(f"PWD={{{credentials.PWD}}}")
+
+            if not getattr(credentials, 'encrypt', False):
+                con_str.append(f"Encrypt={credentials.encrypt}")
 
             con_str_concat = ';'.join(con_str)
             logger.debug(f'Using connection string: {con_str_concat}')
```
Comment (NandanHegde15) [inline dbt/adapters/sqlserver/connections.py:160]: This code displays the credentials in the `dbt logs` without any encryption. So we would prefer this line be commented out and can be used only during debugging purposes.
This fix can be made in a separate PR but wanted to highlight you this scenario.

**Comment by swanderz** [comment]
@mikaelene this PR is all that is needed to enable Azure SQL, which would pretty cool! Like @NandanHegde15 said, we can authenticate and found all the `dbt` commands to be working.

File: dbt/adapters/sqlserver/connections.py [hunk 3]
```diff
@@ -97,16 +126,55 @@ def open(cls, connection):
...
                 con_str.append(f"trusted_connection=yes")
+            elif type_auth == 'sql':
+                con_str.append("Authentication=SqlPassword")
+                con_str.append(f"UID={{{credentials.UID}}}")
+                con_str.append(f"PWD={{{credentials.PWD}}}")
+
+            if not getattr(credentials, 'encrypt', False):
+                con_str.append(f"Encrypt={credentials.encrypt}")
 
             con_str_concat = ';'.join(con_str)
```
Comment (NandanHegde15) [inline dbt/adapters/sqlserver/connections.py:159]: This code displays the credentials in the dbt logs without any encryption. So we would prefer this line be commented out and can be used only during debugging purposes.
This fix can be made in a separate PR but wanted to highlight you this scenario.

File: dbt/adapters/sqlserver/connections.py (hunk 2 above)
Comment (NandanHegde15) [inline dbt/adapters/sqlserver/connections.py:160]: This code displays the credentials in the dbt logs without any encryption. So we would prefer this line be commented out and can be used only during debugging purposes.
This fix can be made in a separate PR but wanted to highlight you this scenario.

**Review by mikaelene — APPROVED** [review]
This looks good @NandanHegde15 ! I'll merge it to master and see if I can try it out before pushing to pipi. Feel free to submit at PR for the credentials debug log as well so we solve that security hole.
